### PR TITLE
feat: scan for vuln

### PR DESCRIPTION
**Pull request by naveensrinivasan**
## Description

- Scans Vulnerabilities for upstream and IronBank
- Uses SBOM and grype to get the Vulns
- The Vulns are reported into GH repo using SARIF

## Related Issue

Fixes #

Relates to #

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Other (security config, docs update, etc)

## Checklist before merging

- [ ] Test, docs, adr added or updated as needed
- [x] [Contributor Guide Steps](https://github.com/defenseunicorns/uds-package-gitlab/blob/main/CONTRIBUTING.md#developer-workflow) followed

### CODE REVIEW

File: .github/workflows/scan-vulnerability.yaml [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+  schedule:
+    - cron: '0 1 * * *'
+  workflow_dispatch: {}
+
+jobs:
+  validate:
+    runs-on: ubuntu-latest
+    name: Scan for vulnerabilities
+    permissions:
+      contents: read
```
Comment (zachariahmiller) [inline .github/workflows/scan-vulnerability.yaml:16]:
```suggestion
      contents: read
```

**Review by zachariahmiller — CHANGES_REQUESTED** [review]

**Review by Racer159 — CHANGES_REQUESTED** [review]
The output here is quite verbose - is it possible to dedupe some of these that are repeated?

https://github.com/defenseunicorns/uds-package-gitlab/security/code-scanning?query=pr%3A98+is%3Aopen

There also seem to be some that are titled as "medium" or "low" but labeled as "critical" or "high"

File: tasks/scanvulnerability.yaml [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+          for flavor in upstream registry1; do
+            uds zarf package create . --flavor="$flavor" --confirm --no-progress -o sbom
+            for file in sbom/*.zst; do
+              uds zarf package inspect "$file" --sbom-out ./sbom --no-progress
+            done
+          done
+          if ! command -v grype &> /dev/null; then
+            echo "Grype could not be found, installing..."
+            curl -sSfL https://raw.githubusercontent.com/anchore/grype/main/install.sh | sh -s -- -b /usr/local/bin
+          fi
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply tasks/scanvulnerability.yaml]: I took care of it.

File: tasks/scanvulnerability.yaml [hunk 3]
```diff
@@ -0,0 +1,25 @@
+---
+tasks:
+  - name: grype-scan-sbom
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply tasks/scanvulnerability.yaml]: 👍

File: .github/workflows/scan-vulnerability.yaml [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+      - name: Checkout
+        uses: actions/checkout@b4ffde65f46336ab88eb53be808477a3936bae11 # v4.1.1
+        with:
+          fetch-depth: 0
+
+      - name: Environment setup # this is required for scanning ironbank images
+        uses: defenseunicorns/uds-common/.github/actions/setup@e2ad99f7caba1b0d08856918db9385a431cfdbca # v0.3.3
+        with:
+          username: ${{ secrets.IRON_BANK_ROBOT_USERNAME }}
+          password: ${{ secrets.IRON_BANK_ROBOT_PASSWORD }}
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply .github/workflows/scan-vulnerability.yaml]: 👍

File: tasks.yaml [hunk 5]
```diff
@@ -7,6 +7,7 @@ includes:
   - pull: https://raw.githubusercontent.com/defenseunicorns/uds-common/v0.3.9/tasks/pull.yaml
   - deploy: https://raw.githubusercontent.com/defenseunicorns/uds-common/v0.3.9/tasks/deploy.yaml
   - setup: https://raw.githubusercontent.com/defenseunicorns/uds-common/v0.3.9/tasks/setup.yaml
+  - vuln-check: ./tasks/scanvulnerability.yaml
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply tasks.yaml]: 👍

File: tasks.yaml [hunk 6]
```diff
@@ -60,3 +61,8 @@ tasks:
       - task: deploy:test-bundle
       - task: test:health-check
       - task: test:ingress
+
+  - name: vuln-scan-package
+    description: Scan the GitLab package for vulnerabilities
+    actions:
+      - task: vuln-check:grype-scan-sbom
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply tasks.yaml]: 👍

File: tasks/scanvulnerability.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+    description: Create a UDS package with configurable flavor and extract the SBOM from all created packages and analyze for vulnerabilities
+    actions:
+      - cmd: |
+          for flavor in upstream registry1; do
+            uds zarf package create . --flavor="$flavor" --confirm --no-progress -o sbom
+            for file in sbom/*.zst; do
+              uds zarf package inspect "$file" --sbom-out ./sbom --no-progress
+            done
+          done
```
[the remark this answers is not in the document]
Comment (naveensrinivasan) [reply tasks/scanvulnerability.yaml]: 👍

**Comment by naveensrinivasan** [comment]
> The output here is quite verbose - is it possible to dedupe some of these that are repeated?
> 
> https://github.com/defenseunicorns/uds-package-gitlab/security/code-scanning?query=pr%3A98+is%3Aopen
> 
> There also seem to be some that are titled as "medium" or "low" but labeled as "critical" or "high"

1. **Image corresponding to the CVE**- all of these CVEs have which image is reporting the CVE. `registry1.dso.mil_ironbank_redhat_ubi_ubi9_9.3:1`
https://github.com/defenseunicorns/uds-package-gitlab/security/code-scanning/10337

2. **Duplicate issue** - All the zlib issues are from different containers.https://github.com/defenseunicorns/uds-package-gitlab/security/code-scanning?query=pr%3A98+is%3Aopen+zlib

3. **Critical Issue having low in the title** - The https://github.com/defenseunicorns/uds-package-gitlab/security/code-scanning/10212 is low because of the probability of exploiting the Vulnerability, and that is the reason low with Vuln being still critical.

Some help from ChatGPT for `low`
TL;DR:
- Vulnerability severity is based on potential impact, exploit complexity, mitigations, impact, user interaction, exploitation environment, patch availability, and historical context.
- Mitigating complexity, mitigations, limited impact, user interaction, and restricted exploitation environment can lower severity.
- Prompt patches and historical context also influence severity classification.
- The process ensures the effective allocation of resources to address pressing security concerns.

**Comment by Racer159** [comment]
closing this do to becoming stale (we can reopen / readdress later)
